core/src/services/huggingface/backend.rs

```diff
@@ -0,0 +1,348 @@
+        if !repo_id.is_empty() {
+            self.config.repo_id = Some(repo_id.to_string());
+        }
+        self
+    }
+
+    /// Set revision of this backend. Default is main.
+    ///
+    /// Revision can be a branch name or a commit hash.
+    ///
+    /// For example, revision can be:
+    /// - main
+    /// - 1d0c4eb
+    pub fn revision(&mut self, revision: &str) -> &mut Self {
+        if !revision.is_empty() {
+            self.config.revision = Some(revision.to_string());
+        }
+        self
+    }
+
+    /// Set root of this backend.
+    ///
+    /// All operations will happen under this root.
+    pub fn root(&mut self, root: &str) -> &mut Self {
+        if !root.is_empty() {
+            self.config.root = Some(root.to_string());
+        }
+        self
+    }
+
+    /// Set the token of this backend.
+    ///
+    /// This is optional.
+    pub fn token(&mut self, token: &str) -> &mut Self {
+        if !token.is_empty() {
+            self.config.token = Some(token.to_string());
+        }
+        self
+    }
+}
+
+impl Builder for HuggingfaceBuilder {
+    const SCHEME: Scheme = Scheme::Huggingface;
+    type Accessor = HuggingfaceBackend;
+
+    fn from_map(map: HashMap<String, String>) -> Self {
+        let mut builder = HuggingfaceBuilder::default();
+
+        map.get("repo_type").map(|v| builder.repo_type(v));
+        map.get("repo_id").map(|v| builder.repo_id(v));
+        map.get("revision").map(|v| builder.revision(v));
+        map.get("root").map(|v| builder.root(v));
+        map.get("token").map(|v| builder.token(v));
+
+        builder
+    }
+
+    /// Build a HuggingfaceBackend.
+    fn build(&mut self) -> Result<Self::Accessor> {
+        debug!("backend build started: {:?}", &self);
+
+        let repo_type = match self.config.repo_type.as_deref() {
+            Some("model") => Ok(RepoType::Model),
+            Some("dataset") => Ok(RepoType::Dataset),
+            Some("space") => Err(Error::new(
+                ErrorKind::ConfigInvalid,
+                "repo type \"space\" is unsupported",
+            )),
+            Some(repo_type) => Err(Error::new(
+                ErrorKind::ConfigInvalid,
+                format!("unknown repo_type: {}", repo_type).as_str(),
+            )
+            .with_operation("Builder::build")
+            .with_context("service", Scheme::Huggingface)),
+            None => Ok(RepoType::Model),
+        }?;
+        debug!("backend use repo_type: {:?}", &repo_type);
+
+        let repo_id = match &self.config.repo_id {
+            Some(repo_id) => Ok(repo_id.clone()),
+            None => Err(Error::new(ErrorKind::ConfigInvalid, "repo_id is empty")
+                .with_operation("Builder::build")
+                .with_context("service", Scheme::Huggingface)),
+        }?;
+        debug!("backend use repo_id: {}", &repo_id);
+
+        let revision = match &self.config.revision {
+            Some(revision) => Ok::<String, Error>(revision.clone()),
```

Comment:
Why we need this?

Reply:
Clippy complains about the absence of specified generic arguments.
```shell
error[E0282]: type annotations needed
   --> core/src/services/huggingface/backend.rs:214:31
    |
214 |             Some(revision) => Ok(revision.clone()),
    |                               ^^ cannot infer type of the type parameter `E` declared on the enum `Result`
    |
help: consider specifying the generic arguments
    |
214 |             Some(revision) => Ok::<std::string::String, E>(revision.clone()),
    |                                 ++++++++++++++++++++++++++

error[E0283]: type annotations needed
   --> core/src/services/huggingface/backend.rs:214:31
    |
214 |             Some(revision) => Ok(revision.clone()),
    |                               ^^ cannot infer type of the type parameter `E` declared on the enum `Result`
215 |             None => Ok("main".to_string()),
216 |         }?;
    |          - type must be known at this point
    |
note: multiple `impl`s satisfying `types::error::Error: std::convert::From<_>` found
   --> core/src/services/etcd/backend.rs:397:1
    |
397 | impl From<EtcdError> for Error {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
   ::: core/src/services/ftp/err.rs:24:1
    |
24  | impl From<FtpError> for Error {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
   ::: core/src/services/redis/backend.rs:449:1
    |
449 | impl From<RedisError> for Error {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
   ::: core/src/services/rocksdb/backend.rs:194:1
    |
194 | impl From<rocksdb::Error> for Error {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
   ::: core/src/services/sftp/error.rs:31:1
    |
31  | impl From<SftpClientError> for Error {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
49  | impl From<SshError> for Error {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
67  | impl From<SftpError> for Error {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: and more `impl`s found in the following crates: `core`:
            - impl<T> std::convert::From<!> for T;
            - impl<T> std::convert::From<T> for T;
    = note: required for `std::result::Result<services::huggingface::backend::HuggingfaceBackend, types::error::Error>` to implement `std::ops::FromResidual<std::result::Result<std::convert::Infallible, _>>`
help: consider specifying the generic arguments
    |
214 |             Some(revision) => Ok::<std::string::String, E>(revision.clone()),
    |                                 ++++++++++++++++++++++++++
```

I'll look into this problem tomorrow.